src/index.ts

```diff
@@ -0,0 +1,283 @@
+import {
+  HTTPError,
+  MaxRedirectsError,
+  ParseError,
+  RequestError,
+  TimeoutError,
+  UnsupportedProtocolError,
+  type TimeoutEvent,
+} from "./errors";
+export * from "./errors";
+import {
+  requestDuration,
+  requestsCount,
+  requestStageDuration,
+} from "./metrics";
+import { parseUrlTemplate } from "./url-template";
+export { registerMetrics } from "./metrics";
+
+interface Headers {
+  [header: string]: number | string | string[] | undefined;
+}
+
+interface Options {
+  baseUrl: string;
+  headers: Headers;
+  timeout?: number;
+  maxRedirects?: number;
+}
+
+interface Timings {
+  start?: number;
+  socket?: number;
+  lookup?: number;
+  connect?: number;
+  upload?: number;
+  response?: number;
+  end?: number;
+  error?: number;
+  phases: {
+    wait?: number;
+    dns?: number;
+    tcp?: number;
+    request?: number;
+    firstByte?: number;
+    download?: number;
+    total?: number;
+  };
+}
+
+interface Result {
+  timings?: Timings;
+  statusCode?: number;
+  body?: unknown;
+}
+
+type Method =
+  | "GET"
+  | "PUT"
+  | "POST"
+  | "PATCH"
+  | "DELETE"
+  | "HEAD"
+  | "OPTIONS";
+
+export class HTTPClient {
+  baseUrl: string;
+  headers: Headers;
+  timeout: number;
+  maxRedirects: number;
+
+  constructor(opts: Options) {
+    const {
+      baseUrl,
+      headers,
+      timeout = 10000,
+      maxRedirects = 5,
+    } = opts;
+
+    // Validate protocol
+    const supportedProtocols = ["http:", "https:"];
+    const url = new URL(baseUrl);
+    if (!supportedProtocols.includes(url.protocol)) {
+      throw new UnsupportedProtocolError(
+        `Unsupported protocol: ${baseUrl}`
+      );
+    }
+
+    this.baseUrl = baseUrl;
+    this.headers = headers;
+    this.timeout = timeout;
+    this.maxRedirects = maxRedirects;
+  }
+
+  async request(
+    method: Method,
+    pathTemplate: string,
+    params = {},
+    body?: unknown,
+    options?: RequestInit
+  ): Promise<Result> {
+    const path = parseUrlTemplate(pathTemplate);
+    const url = new URL(path.expand(params), this.baseUrl);
+
+    const startTime = performance.now();
+
+    const defaultOptions: RequestInit = {
+      method: method.toUpperCase(),
+      headers: {
+        "Content-Type": "application/json",
+        ...this.headers,
+      },
+      body: body ? JSON.stringify(body) : undefined,
+      redirect: "follow",
+    };
+
+    const fetchOptions = { ...defaultOptions, ...options };
+
+    try {
+      const controller = new AbortController();
+      const timeoutId = setTimeout(() => {
+        controller.abort();
+      }, this.timeout);
+
+      fetchOptions.signal = controller.signal;
+
+      const response = await fetch(url.toString(), fetchOptions);
```

Comment:
I believe you can use the built in timeout

  const res = await fetch(url, { signal: AbortSignal.timeout(5000) });